contracts/vesting/src/storage.rs

```diff
@@ -0,0 +1,182 @@
+use curve::{Curve, SaturatingLinear};
+use soroban_sdk::{
+    contracttype, log, panic_with_error, Address, ConversionError, Env, String, TryFromVal, Val,
+};
+
+use crate::error::ContractError;
+
+impl TryFromVal<Env, DataKey> for Val {
+    type Error = ConversionError;
+
+    fn try_from_val(_env: &Env, v: &DataKey) -> Result<Self, Self::Error> {
+        Ok((*v as u32).into())
+    }
+}
+
+#[derive(Clone, Copy)]
+#[repr(u32)]
+pub enum DataKey {
+    Admin = 1,
+    Config = 2,
+    Minter = 3,
+    Whitelist = 4,
+    VestingTokenInfo = 5,
+    MaxVestingComplexity = 6,
+}
+
+#[contracttype]
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct VestingTokenInfo {
+    pub name: String,
+    pub symbol: String,
+    pub decimals: u32,
+    pub address: Address,
+    pub total_supply: i128,
+}
+
+#[contracttype]
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct VestingBalance {
+    pub address: Address,
+    pub balance: i128,
+    pub distribution_info: DistributionInfo,
+}
+
+#[contracttype]
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct MinterInfo {
+    pub address: Address,
+    pub mint_cap: u128, // use this
+}
+
+#[contracttype]
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct DistributionInfo {
+    pub start_timestamp: u64,
+    pub min_value_at_start: u128,
+    pub end_timestamp: u64,
+    pub max_value_at_end: u128,
+}
+
+#[contracttype]
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct VestingInfo {
+    pub amount: i128,
+    pub distribution_info: DistributionInfo,
+}
+
+impl MinterInfo {
+    pub fn get_curve(&self) -> Curve {
+        Curve::Constant(self.mint_cap)
+    }
+}
+
+impl DistributionInfo {
+    pub fn get_curve(&self) -> Curve {
+        Curve::SaturatingLinear(SaturatingLinear {
+            min_x: self.start_timestamp,
+            min_y: self.min_value_at_start,
+            max_x: self.end_timestamp,
+            max_y: self.max_value_at_end,
+        })
+    }
```

Comment:
Following on my previous comment.

```suggestion
    pub fn get_curve(&self) -> Curve {
        Curve::SaturatingLinear(SaturatingLinear {
            min_x: self.start_timestamp,
            min_y: self.amount,
            max_x: self.end_timestamp,
            max_y: 0u128,
        })
    }
```